diff --git a/node_modules/@whiskeysockets/baileys/.bun-tag-6bc22c37f6067614 b/.bun-tag-6bc22c37f6067614
new file mode 100644
index 0000000000000000000000000000000000000000..e69de29bb2d1d6434b8b29ae775ad8c2e48c5391
diff --git a/node_modules/@whiskeysockets/baileys/.bun-tag-ba297a3a96c78f79 b/.bun-tag-ba297a3a96c78f79
new file mode 100644
index 0000000000000000000000000000000000000000..e69de29bb2d1d6434b8b29ae775ad8c2e48c5391
diff --git a/lib/Socket/Client/websocket.js b/lib/Socket/Client/websocket.js
index 977c048b64a6894165e58bca12c80cb13cce5658..692093cd3b6d8a274ccfafd2e6aaa3aad8d02ffe 100644
--- a/lib/Socket/Client/websocket.js
+++ b/lib/Socket/Client/websocket.js
@@ -41,7 +41,7 @@ class WebSocketClient extends types_1.AbstractSocketClient {
             agent: this.config.agent,
         });
         this.socket.setMaxListeners(0);
-        const events = ['close', 'error', 'upgrade', 'message', 'open', 'ping', 'pong', 'unexpected-response'];
+        const events = ['close', 'error', 'message', 'open', 'ping', 'pong'];
         for (const event of events) {
             (_b = this.socket) === null || _b === void 0 ? void 0 : _b.on(event, (...args) => this.emit(event, ...args));
         }
diff --git a/lib/Socket/socket.js b/lib/Socket/socket.js
index 9b39302799cbd768f46d3a4bd6cd7a50ed495e38..0a82c0810063d117d298acfab08a68034ef6ea48 100644
--- a/lib/Socket/socket.js
+++ b/lib/Socket/socket.js
@@ -80,6 +80,9 @@ const makeSocket = (config) => {
     };
     /** log & process any unexpected errors */
     const onUnexpectedError = (err, msg) => {
+        if (!ws.isOpen) {
+            return;
+        }
         logger.error({ err }, `unexpected error in '${msg}'`);
     };
     /** await the next incoming message */
diff --git a/lib/Types/Message.d.ts b/lib/Types/Message.d.ts
index e04823e26465bacb2e672c57c7720bef97ac69e5..79036283e1eba939e7b381e055c4f9280f043766 100644
--- a/lib/Types/Message.d.ts
+++ b/lib/Types/Message.d.ts
@@ -7,11 +7,20 @@ import { BinaryNode } from '../WABinary';
 import type { GroupMetadata } from './GroupMetadata';
 import { CacheStore } from './Socket';
 export { proto as WAProto };
-export type WAMessage = proto.IWebMessageInfo;
+export type WAMessage = proto.IWebMessageInfo & {
+    key: WAMessageKey;
+};
 export type WAMessageContent = proto.IMessage;
 export type WAContactMessage = proto.Message.IContactMessage;
 export type WAContactsArrayMessage = proto.Message.IContactsArrayMessage;
-export type WAMessageKey = proto.IMessageKey;
+export type WAMessageKey = proto.IMessageKey & {
+    senderLid?: string;
+    server_id?: string;
+    senderPn?: string;
+    participantLid?: string;
+    participantPn?: string;
+    isViewOnce?: boolean;
+};
 export type WATextMessage = proto.Message.IExtendedTextMessage;
 export type WAContextInfo = proto.IContextInfo;
 export type WALocationMessage = proto.Message.ILocationMessage;
@@ -107,7 +116,7 @@ export type AnyMediaMessageContent = (({
     isAnimated?: boolean;
 } & WithDimensions) | ({
     document: WAMediaUpload;
-    mimetype: string;
+    mimetype?: string;
     fileName?: string;
     caption?: string;
 } & Contextable)) & {
@@ -211,6 +220,7 @@ export type MiscMessageGenerationOptions = MinimalRelayOptions & {
     font?: number;
     /** if it is broadcast */
     broadcast?: boolean;
+    waveformProxy?: Buffer | null;
 };
 export type MessageGenerationOptionsFromContent = MiscMessageGenerationOptions & {
     userJid: string;
diff --git a/lib/Utils/auth-utils.js b/lib/Utils/auth-utils.js
index 7a6127e952e87a794e516b977c9a74bd0cf5526d..9d04e9141e2ff467d9b7b740d7b15cb3ef75f2bc 100644
--- a/lib/Utils/auth-utils.js
+++ b/lib/Utils/auth-utils.js
@@ -124,6 +124,16 @@ const addTransactionCapability = (state, logger, { maxCommitRetries, delayBetwee
                 return state.set(data);
             }
         },
+        clear: () => {
+            if (isInTransaction()) {
+                logger.trace('clearing cache in transaction');
+                transactionCache = {};
+                mutations = {};
+            }
+            else {
+                return state.clear();
+            }
+        },
         isInTransaction,
         async transaction(work) {
             let result;
diff --git a/lib/Utils/decode-wa-message.js b/lib/Utils/decode-wa-message.js
index b6b7becd2c1a48c3ebe3ee11da7105f02c248e69..7eede5838fb3b08b05797a17c3e3e2ba2a220cfa 100644
--- a/lib/Utils/decode-wa-message.js
+++ b/lib/Utils/decode-wa-message.js
@@ -28,7 +28,7 @@ exports.NACK_REASONS = {
  * @note this will only parse the message, not decrypt it
  */
 function decodeMessageNode(stanza, meId, meLid) {
-    var _a;
+    var _a, _b, _c, _d, _e;
     let msgType;
     let chatId;
     let author;
@@ -87,7 +87,11 @@ function decodeMessageNode(stanza, meId, meLid) {
         remoteJid: chatId,
         fromMe,
         id: msgId,
-        participant
+        senderLid: (_b = stanza === null || stanza === void 0 ? void 0 : stanza.attrs) === null || _b === void 0 ? void 0 : _b.sender_lid,
+        senderPn: (_c = stanza === null || stanza === void 0 ? void 0 : stanza.attrs) === null || _c === void 0 ? void 0 : _c.sender_pn,
+        participant,
+        participantPn: (_d = stanza === null || stanza === void 0 ? void 0 : stanza.attrs) === null || _d === void 0 ? void 0 : _d.participant_pn,
+        participantLid: (_e = stanza === null || stanza === void 0 ? void 0 : stanza.attrs) === null || _e === void 0 ? void 0 : _e.participant_lid
     };
     const fullMessage = {
         key,
diff --git a/lib/Utils/messages-media.js b/lib/Utils/messages-media.js
index 2251c0019df0755d3f65c81ea3b48e963c33cdad..a6033b3463e0a9f5d44df96d609758882114d3a6 100644
--- a/lib/Utils/messages-media.js
+++ b/lib/Utils/messages-media.js
@@ -422,7 +422,11 @@ const toSmallestChunkSize = (num) => {
 const getUrlFromDirectPath = (directPath) => `https://${DEF_HOST}${directPath}`;
 exports.getUrlFromDirectPath = getUrlFromDirectPath;
 const downloadContentFromMessage = async ({ mediaKey, directPath, url }, type, opts = {}) => {
-    const downloadUrl = url || (0, exports.getUrlFromDirectPath)(directPath);
+    const isValidMediaUrl = url === null || url === void 0 ? void 0 : url.startsWith('https://mmg.whatsapp.net/');
+    const downloadUrl = isValidMediaUrl ? url : (0, exports.getUrlFromDirectPath)(directPath);
+    if (!downloadUrl) {
+        throw new boom_1.Boom('No valid media URL or directPath present in message', { statusCode: 400 });
+    }
     const keys = await getMediaKeys(mediaKey, type);
     return (0, exports.downloadEncryptedContent)(downloadUrl, keys, opts);
 };
diff --git a/lib/Utils/messages.js b/lib/Utils/messages.js
index 34d6e7d3d28dc0c5fc2a6ba7bf779d1df4f055d7..e2a97f54224a0a5c517cbf4336ab9ba16c59c0b9 100644
--- a/lib/Utils/messages.js
+++ b/lib/Utils/messages.js
@@ -141,7 +141,7 @@ const prepareWAMessageMedia = async (message, options) => {
                     logger === null || logger === void 0 ? void 0 : logger.debug('computed audio duration');
                 }
                 if (requiresWaveformProcessing) {
-                    uploadData.waveform = await (0, messages_media_1.getAudioWaveform)(originalFilePath, logger);
+                    uploadData.waveform = await (0, messages_media_1.getAudioWaveform)(options.waveformProxy || originalFilePath, logger);
                     logger === null || logger === void 0 ? void 0 : logger.debug('processed waveform');
                 }
                 if (requiresAudioBackground) {
